src/lib/tls/msg_server_hello.cpp

```diff
@@ -750,6 +772,34 @@ Server_Hello_13::Server_Hello_13(const Client_Hello_13& ch,
    if(key_exchange_group.has_value())
       { m_data->extensions.add(new Key_Share(key_exchange_group.value(), cb, rng)); }
 
+   auto& ch_exts = ch.extensions();
+
+   if(ch_exts.has<PSK>())
+      {
+      const auto cs = Ciphersuite::by_id(m_data->ciphersuite);
+      BOTAN_ASSERT_NOMSG(cs);
+
+      // RFC 8446 4.2.9
+      //    A client MUST provide a "psk_key_exchange_modes" extension if it
+      //    offers a "pre_shared_key" extension.
+      //
+      // Note: Client_Hello_13 constructor already performed a graceful check.
+      const auto psk_modes = ch_exts.get<PSK_Key_Exchange_Modes>();
+      BOTAN_ASSERT_NONNULL(psk_modes);
+
+      // TODO: also support non-DHE PSK Key-Exchange mode
+      if(value_exists(psk_modes->modes(), PSK_Key_Exchange_Mode::PSK_DHE_KE))
```

Comment:
The comment is misleading, sorry. What I mean is that we should (at some point) also support a PSK-based exchange that does not perform an additional asymmetric ephemeral key exchange. I'll change that.

For reference: TLS 1.3 allow three key exchange modes:

* Ephemeral key only (the common "full" handshake)
* Ephemeral key + PSK (what we currently support for session resumption)
* PSK only (what this TODO requests to be supported) ... this would be a fully symmetric TLS handshake solely based on a PSK (or previous session) (with weaker forward secrecy guarantees)